include/CLI/Set.hpp

```diff
@@ -0,0 +1,115 @@
+#pragma once
+
+// Distributed under the 3-Clause BSD License.  See accompanying
+// file LICENSE or https://github.com/CLIUtils/CLI11 for details.
+
+#include <initializer_list>
+#include <memory>
+#include <string>
+#include <vector>
+
+// CLI Library includes
+#include "CLI/StringTools.hpp"
+
+namespace CLI {
+
+class Set {
+  protected:
+    using storage_type = std::vector<std::string>;
+    std::shared_ptr<storage_type> items_{std::make_shared<storage_type>()};
```

Comment:
The question of how to enable mutable sets is going to be the tricky part here,  especially how to make it not unexpectedly mutable.

Reply:
Sets are always mutable. Not sure why you would add a Set, then change it and not expect it to change. If you change it, it changes in the App. The thing this fixes is you can’t segfault by deleting the set.

Reply:
I think my hesitancy is that when you see something passed by value it is usually copying(or moving) into it and it is then isolated from any changes that might be made at a higher level.    The shared_ptrs take care of lifetime issues with this, but it seems a little non-intuitive that any modifications to the set effect all previous uses of that set, even though it was passed by value.  

For example I have 1 use case where there are set of 5 different modes for one variable,  another has those 5 modes but adds an additional one.    Seeing sets are passed by value I might naively just add an element to the set and give it to the second option.  which would allow a hard to detect invalid set member in the first.  At the very least I am concerned that this will lead to some hard to detect and frustrating bugs on the part of users.  

I am actually not entirely sure what the best approach might be,  maybe take by pointer,  make a copy with the shared state,  then provide a deep copy function? .

Reply:
Since CLI11 takes references as the main way of setting values, I felt that was pretty reasonable. However, since "mutability" is probably not as commonly needed, and should be clear to the user, I'm thinking that simplifying set and then taking two signatures, the set or a `shared_ptr` to a set, might be a good idea. The shared pointer version would clearly be mutable.